### PR TITLE
Grid outage powerflow calculations

### PR DESCRIPTION
**Pull request by brtietz**
Adds new variables to lib battery powerflow and _partially_ connects them to the battery cmods.

New inputs:
  - isOutageStep
  - powerInterconnectionLimit
  - powerCurtailmentLimit
  - powerCritLoad

New outputs:
  - powerInterconnectionLoss
  - powerCritLoadUnmet

powerInterconnectionLoss calculates the power lost due to interconnection limits, curtailment, or a grid outage. It's not currently possible to specify a grid outage in the GUI (only the tests), but the behavior of the variable can be seen by specifying the interconnection limit to zero.

This is the first of several PRs over the next few weeks for the Puerto Rico project. Future PRs will include:
 - Connecting critical load from the cmod level to this code, possibly including a new escalation variable
 - Refactoring the resilience runner such that it uses this code and calculates losses properly (see https://github.com/NREL/ssc/issues/590)
 - Updating dispatch to respond to an outage (utilizing portions of the resilience runner for simulation steps)

Some of this work builds towards https://github.com/NREL/SAM/issues/181 - since the dispatch classes will now have access towards the numbers they need.

### CODE REVIEW

**Comment by sjanzou** [comment]
"powerInterconnectionLoss calculates the power lost due to interconnection limits, curtailment, or a grid outage. It's not currently possible to specify a grid outage in the GUI (only the tests), but the behavior of the variable can be seen by specifying the interconnection limit to zero." 
This statement seems to conflict with the statement on the Grid Limits page
image

I understand that if the grid is down, then no exporting is possible but I think that the text on the UI should be updated for grid outages... 
@cpaulgilman, what do you think?

File: ssc/cmod_pvsamv1.cpp [hunk 1]
```diff
@@ -2337,8 +2338,14 @@ void cm_pvsamv1::exec()
                 }
 			}
 			// Update battery with final gen to compute grid power
-			if (en_batt)
-				batt->update_grid_power(*this, PVSystem->p_systemACPower[idx], p_load_full[idx], idx);
+            if (en_batt) {
+                if (batt->is_outage_step(idx % nrec)) {
+                    batt->update_grid_power(*this, PVSystem->p_systemACPower[idx], p_crit_load_in[idx % nrec], idx);
```
Comment (sjanzou) [inline ssc/cmod_pvsamv1.cpp:2343]: How dows this work? If the grid is out, how is there any grid power to update?

File: ssc/common.cpp [hunk 2]
```diff
@@ -519,7 +519,10 @@ var_info vtab_grid_curtailment[] = {
 	/*   VARTYPE           DATATYPE         NAME                               LABEL                                       UNITS     META                                     GROUP                 REQUIRED_IF                 CONSTRAINTS                      UI_HINTS*/
 
 		{ SSC_INPUT,        SSC_ARRAY,       "grid_curtailment",              "Grid curtailment as energy delivery limit (first year)",              "MW",    "",                                     "GridLimits",      "?",                     "",                "" },
-	var_info_invalid };
+        { SSC_INPUT,        SSC_NUMBER,      "enable_interconnection_limit",      "Enable grid interconnection limit",     "0/1",     "Enable a grid interconnection limit",   "GridLimits",        "","",                           "" },
```
Comment (sjanzou) [inline ssc/common.cpp:522]: Okay, this takes care of the systems without storage, please disregard previous statement removing from cmod_grid.

**Review by sjanzou — CHANGES_REQUESTED** [review]
Everything seems okay, I would still like to go through with the tests and merge in the PVSmoothing changes so I do not have to update all the calls to dispatch_t with the additional argument added to the base class ;-)

File: ssc/cmod_battwatts.cpp [hunk 3]
```diff
@@ -215,6 +215,13 @@ battwatts_create(size_t n_recs, size_t n_years, int chem, int meter_pos, double
     batt_vars->batt_minimum_SOC = 15.;
     batt_vars->batt_minimum_modetime = 10;
 
+    // Interconnection and curtailment
+    batt_vars->gridCurtailmentLifetime_MW = curtailment_limit;
+    batt_vars->grid_interconnection_limit_kW = interconnection_limit;
+    if (interconnection_limit < 1e+38) {
+        batt_vars->enable_interconnection_limit = true;
```
[the remark this answers is not in the document]
Comment (brtietz) [reply ssc/cmod_battwatts.cpp:222]: This particular chunk of code doesn't have access to the vartable, so it has to interpret the state of the boolean from the double value. The ssc vartable value you're thinking of is checked in cm_battwatts::setup_variables. I thought this was preferable to passing in two inputs to this function, but welcome other opinions (such as passing in the vartable).

File: ssc/cmod_grid.cpp [hunk 4]
```diff
@@ -34,8 +34,6 @@ var_info vtab_grid_input[] = {
 	// simulation inputs
 	{ SSC_INPUT,        SSC_NUMBER,      "system_use_lifetime_output",        "Lifetime simulation",                   "0/1",     "0=SingleYearRepeated,1=RunEveryYear",   "Lifetime",        "?=0",                   "BOOLEAN",                          "" },
 	{ SSC_INPUT,        SSC_NUMBER,      "analysis_period",                   "Lifetime analysis period",              "years",   "The number of years in the simulation", "Lifetime",        "system_use_lifetime_output=1","",                           "" },
-	{ SSC_INPUT,        SSC_NUMBER,      "enable_interconnection_limit",      "Enable grid interconnection limit",     "0/1",     "Enable a grid interconnection limit",   "GridLimits",        "","",                           "" },
```
[the remark this answers is not in the document]
Comment (brtietz) [reply ssc/cmod_grid.cpp:37]: These were moved to common.cpp so the battery code can access them consistently.

File: ssc/cmod_pvsamv1.cpp (hunk 1 above)
Comment (brtietz) [reply ssc/cmod_pvsamv1.cpp:2343]: This is responsible for updating the interconnection loss variable, given the critical load and the updated PV generation, given the set of AC losses that weren't available when the battery is first run.

**Comment by brtietz** [comment]
> 
> 
> Everything seems okay, I would still like to go through with the tests and merge in the PVSmoothing changes so I do not have to update all the calls to dispatch_t with the additional argument added to the base class ;-)

I'm fine to hold this until PV smoothing is merged in and take care of the merge conflicts on this branch, if we think we can get PV smoothing merged in within the next week or so.

**Comment by cpaulgilman** [comment]
> 
> 
> "powerInterconnectionLoss calculates the power lost due to interconnection limits, curtailment, or a grid outage. It's not currently possible to specify a grid outage in the GUI (only the tests), but the behavior of the variable can be seen by specifying the interconnection limit to zero."
> This statement seems to conflict with the statement on the Grid Limits page
> image
> 
> I understand that if the grid is down, then no exporting is possible but I think that the text on the UI should be updated for grid outages...
> @cpaulgilman, what do you think?

I can update UI (and Help) descriptions like these after the code is implemented. The "interconnection loss" includes all power "lost" at the interconnection point, i.e., not delivered to the grid because of the interconnection limit, curtailment, or a grid outage.

File: test/shared_test/lib_battery_powerflow_test.cpp [hunk 5]
```diff
@@ -1593,3 +1995,307 @@ TEST_F(BatteryPowerFlowTest_lib_battery_powerflow, TestDCConnected)
...
+    EXPECT_NEAR(m_batteryPower->powerBatteryToLoad, 0, error);
+    EXPECT_NEAR(m_batteryPower->powerConversionLoss, 2.39, error);
+    EXPECT_NEAR(m_batteryPower->powerSystemLoss, 1.0, error);
+    EXPECT_NEAR(m_batteryPower->powerCritLoadUnmet, 48.39, error);
+    EXPECT_NEAR(m_batteryPower->powerInterconnectionLoss, 0.0, error);
+
+    check_net_flows("6th case");
+
+    // excess discharging causes interconnection loss
+    m_batteryPower->powerBatteryDC = 50;
```
Comment (dguittet) [inline test/shared_test/lib_battery_powerflow_test.cpp:2211]: here the power of the battery was curtailed and interconnectionloss is 0, similar to an above case around line 847

File: test/shared_test/lib_battery_powerflow_test.cpp [hunk 6]
```diff
@@ -1593,3 +1995,307 @@ TEST_F(BatteryPowerFlowTest_lib_battery_powerflow, TestDCConnected)
...
+    EXPECT_NEAR(m_batteryPower->powerGridToBattery, 0, error);
+    EXPECT_NEAR(m_batteryPower->powerGridToLoad, 0, error);
+    EXPECT_NEAR(m_batteryPower->powerSystemToGrid, 0, error);
+    EXPECT_NEAR(m_batteryPower->powerBatteryToLoad, 26.32, error);
+    EXPECT_NEAR(m_batteryPower->powerConversionLoss, 3.39, error);
+    EXPECT_NEAR(m_batteryPower->powerSystemLoss, 1.0, error);
+    EXPECT_NEAR(m_batteryPower->powerCritLoadUnmet, 0, error);
+    EXPECT_NEAR(m_batteryPower->powerInterconnectionLoss, 0.0, error);
+
+    check_net_flows("7th case");
```
Comment (dguittet) [inline test/shared_test/lib_battery_powerflow_test.cpp:2226]: seems like the battery charging will not be reduced to reduce unmet load, and the battery will be curtailed to reduce interconnection loss. Is this always true?

**Review by dguittet — APPROVED** [review]
Nice work! LGTM. 

I think some updated comments, and updating this PR's info and the things in Help to make clear what power flows are changed and how, by an outage vs grid limits, etc (such as asked by my comments).

File: ssc/cmod_battery.cpp [hunk 7]
```diff
@@ -1534,13 +1598,25 @@ void battstor::metrics()
...
-// function needed to correctly calculate P_grid to to additional losses in P_gen post battery like wiring, curtailment, availablity
+// function needed to correctly calculate P_grid due to additional losses in P_gen post battery like wiring, curtailment, availablity
 void battstor::update_grid_power(compute_module&, double P_gen_ac, double P_load_ac, size_t index_replace)
 {
-    double P_grid = P_gen_ac - P_load_ac;
+    double P_interconnection_loss = outInterconnectionLoss[index_replace];
+    double P_grid = P_gen_ac - P_load_ac - P_interconnection_loss;
+    if (P_grid < 0 && P_gen_ac > P_load_ac) {
+        outInterconnectionLoss[index_replace] += P_grid;
+        P_grid = 0;
```
[the remark this answers is not in the document]
Comment (brtietz) [reply ssc/cmod_battery.cpp:1608]: This is intended to correct the interconnection loss given the updated PV losses passed into this function. If gen is higher then load, but the calculation on 1605 is negative, that means the interconnection loss is now too high. I'll add a comment to this effect.

[the remark this answers is not in the document]
Comment (brtietz) [reply ssc/cmod_battery.cpp:1608]: Oh, I understand now - this code is actually missing the case where we need to adjust _both_ the grid variable and the interconnection variable due to losses. Right now export smaller than the interconnection loss won't get reported in P_grid.

File: test/shared_test/lib_battery_powerflow_test.cpp (hunk 6 above)
Comment (brtietz) [reply test/shared_test/lib_battery_powerflow_test.cpp:2226]: Yes to the first part - I view that as the job of dispatch code, not the powerflow code.

For the second part - that's only true during an outage. For an interconnection limit, PV's export to grid is curtailed first, followed by fuel cell, followed by battery discharging. However, P_battery_ac is not yet reduced for an interconnection limit, so there's work for dispatch to do there too.

Comment (sjanzou) [reply test/shared_test/lib_battery_powerflow_test.cpp:2226]: @brtietz, do we need to funnel in the grid interconnection limits to the dispatch code to account for additional adjustments to P_battery_ac?

Comment (brtietz) [reply test/shared_test/lib_battery_powerflow_test.cpp:2226]: Yes, that is planned for a subsequent pull request.

File: ssc/cmod_pvsamv1.cpp [hunk 8]
```diff
@@ -684,6 +684,10 @@ static var_info _cm_vtab_pvsamv1[] = {
 
         { SSC_OUTPUT,        SSC_ARRAY,     "ac_transmission_loss",                   "Transmission loss",                                     "kW", "",    "Time Series (Transmission)",                 "",                     "",                   "" },
 
+        // Post batt AC losses - record so the powerflows from PV and batt to grid add up properly
+        { SSC_OUTPUT,        SSC_ARRAY,     "ac_perf_adj_loss",                       "AC performance adjustment loss",                             "kW", "",    "Time Series (AC Loss)",                 "",                     "",                   "" },
```
Comment (sjanzou) [inline ssc/cmod_pvsamv1.cpp:688]: @cpaulgilman, do you like the "friendly" names here?

**Review by sjanzou — APPROVED** [review]
Thanks for updating the pv smoothing library. Did Paul sign off on the new output names? LGTM and all tests passing!

**Review by dguittet — APPROVED** [review]
